### ISSUE TITLE
Multiple false positives in simple example

### ISSUE THREAD

**Issue by walles** [YorickPeterse/ruby-lint#142]
Put this in math_test.rb:
```ruby
require 'test/unit'

class MathTest < Test::Unit::TestCase
  def test_addition
    assert_equal(4, 2 + 2)
  end
end
```

Here's what I get when I try to ruby-lint this:
```
/t/e/test $ ruby-lint math_test.rb 
math_test.rb: error: line 3, column 18: undefined constant Test::Unit::TestCase
math_test.rb: error: line 3, column 18: undefined constant Test::Unit
math_test.rb: error: line 3, column 18: undefined constant Test
math_test.rb: error: line 5, column 5: undefined method assert_equal
```

I would have expected this program to pass without warnings.

Am I doing something wrong here? What in that case?

My environment is:
* Mac OS X 10.10.2
* ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
* ruby-lint v2.0.3 on ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]

**Comment by YorickPeterse** [comment]
ruby-lint doesn't know about `Test::Unit` & friends, this requires extra definitions (https://github.com/YorickPeterse/ruby-lint/tree/master/lib/ruby-lint/definitions) to be generated.

**Comment by walles** [comment]
If I want to `require` my own code, how can I get ruby-lint to understand that?

**Comment by YorickPeterse** [comment]
ruby-lint tries to figure out where code was defined in a project based on the directories configured using http://code.yorickpeterse.com/ruby-lint/latest/file.configuration.html#directories. This means that if it bumps into a constant `User` it will search these directories for a `user.rb` file defining said class/module. ruby-lint does not follow `require()` calls, some more backstory on this can be found in https://github.com/YorickPeterse/ruby-lint/issues/129#issuecomment-51491887.

**Comment by jawshooah** [comment]
Perhaps it would be prudent to allow the user to specify a mapping of constants to the files where they're defined in `ruby-lint.yml` when they don't exactly match the expected naming scheme.

Maybe something like:
```yaml
constant_definitions:
  Foo: lib/foo/foo_module.rb
  Foo::Bar: lib/foo/bar/bar_module.rb
```

**Comment by walles** [comment]
Sounds reasonable as a complement to the current scheme.

Also, if this could be combined with looking for the config file in...
1. `.`
2. `..`
3. `../..`
4. `../../..`
5 Etc all the way up to /

... then ruby-lint should become much more usable with Emacs Flycheck mode (which is how I attempted to use it).

Thanks /Johan

**Comment by YorickPeterse** [comment]
> Perhaps it would be prudent to allow the user to specify a mapping of constants to the files where they're defined in ruby-lint.yml when they don't exactly match the expected naming scheme.

Requiring users to keep such a mapping up to date isn't something that scales well. If the project structure changes the user has to again update their configuration file. I suggest instead following the common Ruby style guide of naming files after the primary class/module defined in them (so `user.rb` for a User class, etc).

> ... then ruby-lint should become much more usable with Emacs Flycheck mode (which is how I attempted to use it).

Quite a bit of this is discussed in #130. My opinion still is that if Flycheck changes the working directory to the parent directory of a file that is a problem with Flycheck. ruby-ling has an option to specify a custom configuration file, which flycheck could set if needed.

**Comment by dgutov** [comment]
> My opinion still is that if Flycheck changes the working directory to the parent directory of a file that is a problem with Flycheck.

I think we've been over this: in Emacs, the current directory is always the parent directory of the current file, unless specified otherwise. Maybe Flycheck could change the current directory temporarily when calling `ruby-lint`, but that's doing more, not less.

**Comment by YorickPeterse** [comment]
Correct, and I still stand by requiring people to run ruby-lint from the correct directory instead of it traversing directories upwards (or a derivation thereof). The problem as described by OP has nothing to do with this, their problem is that ruby-lint lacks definitions for `Test::Unit` at the moment.

**Comment by YorickPeterse** [comment]
Definitions for test-unit were added in 6bea95313cf887f3d5abbd5ba677d614ea630df9 which has been released as version 2.0.4, this solves the method errors for `Test::Unit` subclasses.
